pages/posts/2017/12-december/140-continuous-integration-at-subvisual.md

```diff
@@ -0,0 +1,116 @@
+intro: "Learn how we do Continuous Integration, and what tools we use at each step of the way"
+---
+
+We're often approached by other companies or developers who want to know more about how we handle our projects' testing and deployment.
+
+Especially since we build and ship multiple times a day and often have multiple developers working in multiple branches simultaneously.
+
+This post will give you the highlights of how all of is done at Subvisual.
+
+We don't claim to have that one methodology to rule them all, but hopefully you can extract something from our process and apply it to your own.
+
+So, off we go...
+
+# Automated builds
+
+Our Continuous Integration process happens mostly through our CI, [Semaphore](https://semaphoreci.com/).
+
+It's where we automate our builds and set up automatic deploys.
+
+![2017-10-29-173955_1204x969_scrot.png](https://s3.eu-west-1.amazonaws.com/subvisual/blog/posts/140/semaphore1.png) 
+
+But let's start with the basics. How do we split our work?
+
+# Git Branches
+
+Our current approach regarding Git is very simple.
+
+I remember when, during my first days at Subvisual, we were experimenting with Git workflows, and were very keen on trying out [a successful Git branching model] and other such workflows. These methods ensure that everything on the `master` branch has gone through some sort of testing pipeline and is ready for production use.
+With time and experience, we ended up noticing how this wasn't working for us. With our focus on rapid development & deploy, we often want to get new features and fixes up quickly. The more steps we have to go through to get something to `master`, the slower it will be.
+
+Nowadays, we dropped the more complex approaches in favor of a simpler one. We have a `master` branch, from which all other branches are created, and to which they are all merged to. That's it.
+
+Branches usually have an owner and are only touched by that owner. But we also have no problem sharing them if it suits the needs. For example, I may start the back-end logic required for a new feature, and then hand over the branch so that a front-end developer and a designer can work on the rest. Or we might pair program on it if we feel the need for more coordination.
+
+Once we merge our changes, Semaphore will take care of checking the new build and, if automatic deploys are enabled, put it live.
+
+Of course, "with great power, comes great responsibility". We don't just merge to `master` whenever we wish to. But instead of a more rigorous git workflow, we rely on a strong automated-testing culture. GitHub and Semaphore also help a great deal with the process as well.
+
+# Code Reviews
+
+We've been doing code reviews and enforcing them as required for all of our projects for a long time now. Our rule is simple: every change that is merged must be approved by at least one person other than yourself. So you either need someone else to review it, or you can pair with someone during development.
```

Comment:
Something like `Every change has to be reviewed by someone else. We usually do pull requests, but if you're pair programming we assume that it was already reviewed by your colleague.`